pkg/upgrade/docker_utils_test.go

```diff
@@ -0,0 +1,236 @@
+package upgrade_test
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/stretchr/testify/require"
+	"github.com/xmtp/xmtpd/pkg/testutils"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+)
+
+const testFlag = "ENABLE_UPGRADE_TESTS"
+
+func skipIfNotEnabled() {
+	if _, isSet := os.LookupEnv(testFlag); !isSet {
+		fmt.Printf("Skipping upgrade test. %s is not set\n", testFlag)
+		os.Exit(0)
+	}
+}
+
+func getScriptPath(scriptName string) string {
+	_, filename, _, _ := runtime.Caller(0)
+	baseDir := filepath.Dir(filename)
+	return filepath.Join(baseDir, scriptName)
+}
+
+func loadEnvFromShell() (map[string]string, error) {
+	scriptPath := getScriptPath("./scripts/load_env.sh")
+	cmd := exec.Command(scriptPath)
+	var outBuf, errBuf bytes.Buffer
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+
+	err := cmd.Run()
+	if err != nil {
+		return nil, fmt.Errorf(
+			"error loading env via shell script: %v\nError: %s",
+			err,
+			errBuf.String(),
+		)
+	}
+
+	envMap := make(map[string]string)
+	scanner := bufio.NewScanner(&outBuf)
+	for scanner.Scan() {
+		line := scanner.Text()
+		parts := strings.SplitN(line, "=", 2)
+		if len(parts) == 2 {
+			envMap[parts[0]] = parts[1]
+		}
+	}
+	return envMap, nil
+}
+
+func expandVars(vars map[string]string) {
+	vars["XMTPD_REPLICATION_ENABLE"] = "true"
+	vars["XMTPD_INDEXER_ENABLE"] = "true"
+
+	dbName := testutils.GetCallerName(3) + "_" + testutils.RandomStringLower(6)
+
+	vars["XMTPD_DB_NAME_OVERRIDE"] = dbName
+}
+
+func convertLocalhost(vars map[string]string) {
+	for varKey, varValue := range vars {
+		if strings.Contains(varValue, "localhost") {
+			vars[varKey] = strings.Replace(varValue, "localhost", "host.docker.internal", -1)
+		}
+	}
+}
+
+func dockerRmc(containerName string) error {
+	killCmd := exec.Command("docker", "rm", containerName)
+	return killCmd.Run()
+}
+
+func dockerKill(containerName string) error {
+	killCmd := exec.Command("docker", "kill", containerName)
+	return killCmd.Run()
+}
+
+func constructVariables(t *testing.T) map[string]string {
+	envVars, err := loadEnvFromShell()
+	require.NoError(t, err)
+	expandVars(envVars)
+	convertLocalhost(envVars)
+
+	return envVars
+}
+
+func streamDockerLogs(containerName string) (chan string, func(), error) {
+	logsCmd := exec.Command("docker", "logs", "-f", containerName)
+	stdoutPipe, err := logsCmd.StdoutPipe()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = logsCmd.Start()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	logChan := make(chan string)
+	go func() {
+		scanner := bufio.NewScanner(stdoutPipe)
+		for scanner.Scan() {
+			logChan <- scanner.Text()
+		}
+		close(logChan)
+	}()
+
+	cancelFunc := func() {
+		_ = logsCmd.Process.Kill()
+	}
+
+	return logChan, cancelFunc, nil
+}
+
+func runContainer(
```

Comment:
I think the docker_utils are correct for a first approach, but we could polish and simplify the test by using [testcontainers-go](https://github.com/testcontainers/testcontainers-go?tab=readme-ov-file)